Orange/modelling/fitter.py

```diff
@@ -0,0 +1,94 @@
+from collections import namedtuple
+
+from Orange.base import Learner, Model
+
+LearnerTypes = namedtuple('LearnerTypes', ['classification', 'regression'])
+
+
+class Fitter(Learner):
+    """Handle multiple types of target variable with one learner.
+
+    Subclasses of this class serve as a sort of dispatcher. When subclassing,
+    we provide a `LearnerTypes` instance which contain actual learner classes
+    that handle appropriate data types. The fitter can then be used on any
+    data and will delegate the work to the appropriate learner.
+
+    If the learners that handle each data type require different parameters,
+    you should pass in all the possible parameters to the fitter. The fitter
+    will then determine which parameters have to passed to individual learners.
+
+    """
+    __fits__ = None
+    __returns__ = Model
+
+    # Constants to indicate what kind of problem we're dealing with
+    CLASSIFICATION, REGRESSION = range(2)
+
+    def __init__(self, *args, preprocessors=None, **kwargs):
+        super().__init__(preprocessors=preprocessors)
+        self.args = args
+        self.kwargs = kwargs
+        # Make sure to pass preprocessor params to individual learners
+        self.kwargs['preprocessors'] = preprocessors
+        self.problem_type = self.CLASSIFICATION
+        self.__regression_learner = self.__classification_learner = None
+
+    def __call__(self, data):
+        # Set the appropriate problem type from the data
+        self.problem_type = self.CLASSIFICATION if \
+            data.domain.has_discrete_class else self.REGRESSION
+        return self.learner(data)
+
+    def __get_kwargs(self, kwargs, problem_type):
+        if problem_type == self.CLASSIFICATION:
+            params = self._get_learner_kwargs(self.__fits__.classification)
+            kwarg_keys = params & set(kwargs.keys())
+            kwargs = {k: kwargs[k] for k in kwarg_keys}
+        elif problem_type == self.REGRESSION:
+            params = self._get_learner_kwargs(self.__fits__.regression)
+            kwarg_keys = params & set(kwargs.keys())
+            kwargs = {k: kwargs[k] for k in kwarg_keys}
+        return kwargs
+
+    @property
+    def classification_learner(self):
+        self.__check_dispatches(self.__fits__)
+        if self.__classification_learner is None:
+            self.__classification_learner = self.__fits__.classification(
+                *self.args,
+                **self.__get_kwargs(self.kwargs, self.CLASSIFICATION))
+        self.__classification_learner.use_default_preprocessors = \
+            self.use_default_preprocessors
+        return self.__classification_learner
+
+    @property
+    def regression_learner(self):
```

Comment:
The two properties can be a single, simple function: `get_learner(problem_type)`